Katello/katello
# Fixes #34880 - auto-create missing distributions and improve errors

#### What are the changes introduced in this pull request?
This PR improves distribution auto-creation and error wording when publications are missing.

#### Considerations taken when implementing this change?
I removed `middleware.use Actions::Middleware::ExecuteIfContentsChanged` from `SaveDistributionReferences` because it was getting in the way of saving references when a distribution is recreated. It seems perfectly safe to run even if no distributions are created and shouldn't waste any noticible time.

#### What are the testing steps for this pull request?
1) Sync a repo and set the mirror policy to 'mirror complete'
2) Destroy the repo's distribution by looking up the distribution reference
```ruby
::Katello::Repository.find(id).backend_service(SmartProxy.pulp_primary).distribution_reference.href
```
Then destroy it via the Pulp API or the Pulp CLI (recommended! https://docs.pulpproject.org/pulp_cli/install/#Katello)
3) Sync the repo and check that the distribution reference changed.
  -> Also check that content is hosted properly.
4) Destroy the repo's publication in Pulp now.  Find the publication with `::Katello::Repository.find(id).publication_href`
5) Add the repo to a content view without filters and try to publish. See the new helpful error.  Skip the failing publish steps and destroy the CVV.
6) Sync the repo normally. Note the new helpful error.  Follow the error's instructions and see that the repo syncs and has properly hosted content.
7) Try publishing the CV again and see that it works.
8) Run steps 1-7 on a new repo with a different mirroring policy ('additive' or 'content only').
